File: Python/CoinFlip.py
from random import *
import time

def roll_dice(rolls):
        for roll in range(rolls):
            print ("Your dice rolled: %d" % randint(1,6))


def longest_streak(flips):
    headStreak = 0
    tailStreak = 0
    longestStreak = 0
    for flip in range(flips):
        coin = randint(0,1)
        if (coin == 0):
            headStreak += 1
            tailStreak = 0
        else:
            tailStreak += 1
            headStreak = 0
        if(headStreak > longestStreak or tailStreak > longestStreak):
            longestStreak += 1

    print(("The longest streak of either heads or tails attained by flipping %d coins is: %d") % (flips, longestStreak))

def coin_flip(streak, trials):
    streakcount = 0
    count = 0
    results = []
    for i in range(trials):
    # Repeat the sequence until the desiginated streak is reached.
        while streakcount < streak:
            coin = randint(1,2)
            count += 1
            if coin == 1:
                streakcount += 1
            else:
                streakcount = 0
        results.append(count)
        count = 0
        streakcount = 0
    average = sum(results)/len(results)
    print (("The average number of coin flips to reach a streak of %d heads in a row with %d trials is: %d") % (streak, trials, average))




def coin_flip_specific_average(streak, average):
    streakcount = 0
    count = 0
    results = []
    mean = 0
    trials = 0
    while (mean != average):
    # Repeat the sequence until the desiginated streak is reached.
        while (streakcount < streak):
            coin = randint(1,2)
            count += 1
            if coin == 1:
                streakcount += 1
            else:
                streakcount = 0
        results.append(count)
        count = 0
        streakcount = 0
        mean = sum(results)/len(results)
        trials += 1
        if trials >= 10000:
            print ("Unfortunatley, the average %d could not be reached in %d trials" % (average, trials))
            return
    print (("The number of trials to reach an average number of %d coin flips to reach %d consecutive heads is: %d") % (average, streak, trials))


def die_roll_side_streak(side, streak):
    if (side < 1 or side > 6):
        print("Error: invalid side")
        return
    streakcount = 0
    count = 0
    # Repeat the sequence until the desiginated streak is reached.
    while streakcount < streak:
        if randint(1,6) == side:
            streakcount += 1
        else:
            streakcount = 0
        count += 1
    print (("The number of dice rolls to reach a streak of %d consecutive %ds in a row is: %d") % (streak, side, count))


def die_roll_streak_any(streak, trials):
    # Values that will correspond to the streak of the number rolled
    one = 0
    two = 0
    three = 0
    four = 0
    five = 0
    six = 0
    count = 0
    results = []
    for trial in range(trials):
        # Repeat the sequence until streak is reached for any number.
        while (one or two or three or four or five or six) < streak:
            dice = randint(1,6)
            #Uncomment the line below if you do not believe this code is legit!
            #print ("The number is: %d" % (die))
            count += 1
            if dice == 1:
                one += 1
                '''All other die number variables must be set to zero as shown below in order to
                reset all other streaks to zero because each roll is independent of each other
                and any number could occur before any other number.'''
                two = 0
                three = 0
                four = 0
                five = 0
                six = 0
            elif dice == 2:
                two += 1
                one = 0
                three = 0
                four = 0
                five = 0
                six = 0
            elif dice == 3:
                three += 1
                one = 0
                two = 0
                four = 0
                five = 0
                six = 0
            elif dice == 4:
                four += 1
                one = 0
                two = 0
                three = 0
                five = 0
                six = 0
            elif dice == 5:
                five += 1
                one = 0
                two = 0
                three = 0
                four = 0
                six = 0
            elif dice == 6:
                six += 1
                one = 0
                two = 0
                three = 0
                four = 0
                five = 0
        results.append(count)
        count = 0
        one = 0
        two = 0
        three = 0
        four = 0
        five = 0
        six = 0
    print (("The average number of dice rolls to reach a streak of %d for any number with %d trials is: %d") % (streak, trials, sum(results)/len(results)))


def coin_flip_streak_any(streak, trials):
    heads = 0
    tails = 0
    count = 0
    results = []
    for trial in range(trials):
        # Repeat the sequence until the desiginated streak is reached.
        while (heads or tails) < streak:
            coin = randint(1,2)
            #Uncomment the next four lines if you do not believe this code is legit!
            '''if coin == 1:
                print ("The coin is heads")
            else:
                print ("The coin is tails")'''
            if coin == 1:
                heads+= 1
                tails = 0
            else:
                heads = 0
                tails += 1
            count += 1
        results.append(count)
        count = 0
        heads = 0
        tails = 0

    print (("The average number of coin flips to reach a streak of %d for either heads or tails with %d trials is: %d") % (streak, trials, sum(results)/len(results)))


def coin_flip_heads_percentage(flips):
        heads = 0
        for i in range(flips):
            if  randint(0,1) == 0:
                heads += 1
        print (("With %d coin tosses, %5.2f percent of the flips landed on heads.") % (flips, float(heads)/float(flips) * 100))

def random_number_of_the_day(min, max):
    print ("Random number of the day from %d to %d: %d" % (min, max, randint(min,max)))


longest_streak(1000000000)
'''roll_dice(2)
coin_flip(10, 1000)
coin_flip_specific_average(10, 2020)
die_roll_side_streak(6, 6)
die_roll_streak_any(5, 100)
coin_flip_streak_any(15, 100)
coin_flip_heads_percentage(12345)
random_number_of_the_day(1, 100)'''

File: CPP/CoinFlip.cpp
#include <iostream>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <vector>
#include <numeric>
#include <iomanip>

using namespace std;

void roll_dice(int rolls){
  srand(time(0));
  for (int roll = 0; roll < rolls; roll ++){
    cout << "Your dice rolled: " << (rand() % 6) + 1 << endl;
  }
}

void coin_flip(int streak, int trials){
  srand(time(0));
  int streakcount = 0;
  int count = 0;
  vector<int> results;
  for (int trial = 0; trial < trials; trial++){
    while (streakcount < streak){
      int coin = rand() % 2;
      count++;
      if (coin == 0){
        streakcount ++;
      }
      else{
        streakcount = 0;
      }
  }
    results.push_back(count);
    count = 0;
    streakcount = 0;
  }
  int sum = accumulate(results.begin(), results.end(), 0);
  int average = sum / results.size();
  cout << "The average number of coin flips to reach a streak of " << streak;
  cout << " heads in a row with " << trials << " trials is: " << average << endl;
}

void coin_flip_specific_average(int streak, int average){
  srand(time(0));
  int streakcount = 0;
  int count = 0;
  vector<int> results;
  int mean = 0;
  int trials = 0;
  int coin;
  while (trials < 10000){
    while (streakcount < streak){
      coin = rand() % 2;
      count++;
      if (coin == 0){
        streakcount++;
      }
      else{
        streakcount = 0;
      }
    }
    results.push_back(count);
    count = 0;
    streakcount = 0;
    int sum = accumulate(results.begin(), results.end(), 0);
    int mean = sum / results.size();
    trials++;
    if (mean - average == 0){
      cout << "The number of trials to reach an average number of " << average;
      cout << " coin flips to reach " << streak << " consecutive heads is: " << trials << endl;
      return;
    }
  }
  cout << "Unfortunatley, the average "<< average << " could not be reached in " << trials << " trials" << endl;

}

void dice_roll_side_streak(int side, int streak){
  if(side < 1 || side > 6){
    cout << "Error: invalid side" << endl;
    return;
  }
  srand(time(0));
  int streakcount = 0;
  int count = 0;
  while (streakcount < streak){
    if ((rand() & 6 + 1) == side){
      streakcount++;
    }
    else{
      streakcount = 0;
    }
    count++;
  }
  cout << "The number of dice rolls to reach a streak of " << streak;
  cout << " consecutive " << side << "s in a row is: " << count << endl;
}

void dice_roll_streak_any(int streak, int trials){
  srand(time(0));
  int count = 0;
  int one = 0;
  int two = 0;
  int three = 0;
  int four = 0;
  int five = 0;
  int six = 0;
  int dice;
  int streakcount = 1;
  vector<int> results;
  for (int trial = 0; trial < trials; trial++){
    while(streakcount < streak){
      dice = (rand() % 6) + 1;
      count++;
      if (dice == 1){
        one++;
        two = 0;
        three = 0;
        four = 0;
        five = 0;
        six = 0;
      }
      else if(dice == 2){
        two++;
        one = 0;
        three = 0;
        four = 0;
        five = 0;
        six = 0;
      }
      else if(dice == 3){
        three++;
        one = 0;
        two = 0;
        four = 0;
        five = 0;
        six = 0;
      }
      else if(dice == 4){
        four++;
        one = 0;
        two = 0;
        three = 0;
        five = 0;
        six = 0;
      }
      else if(dice == 5){
        five++;
        one = 0;
        two = 0;
        three = 0;
        four = 0;
        six = 0;
      }
      else if(dice == 6){
        six++;
        one = 0;
        two = 0;
        three = 0;
        four = 0;
        five = 0;
      }
      if(one > 1 || two > 1|| three > 1|| four > 1|| five > 1|| six > 1){
        streakcount++;
      }
      else{
        streakcount = 1;
      }
    }
    results.push_back(count);
    count = 0;
    streakcount = 1;
    one = 0;
    two = 0;
    three = 0;
    four = 0;
    five = 0;
    six = 0;
  }
  int sum = accumulate(results.begin(), results.end(), 0);
  int average = sum / results.size();
  cout << "The average number of dice rolls to reach a streak of " << streak;
  cout << " for any number with " << trials << " trials is: " << average << endl;
}

void coin_flip_streak_any(int streak, int trials){
  srand(time(0));
  int heads_streak = 0;
  int tails_streak = 0;
  int count = 0;
  int streakcount = 1;
  int coin;
  vector<int> results;
  for (int trial = 0; trial < trials; trial++){
      while (streakcount < streak){
        coin = rand() % 2;
        if (coin == 0){
          heads_streak++;
          tails_streak = 0;
        }
        else{
          heads_streak = 0;
          tails_streak++;
        }
        if (heads_streak > 1 || tails_streak > 1){
          streakcount++;
        }
        else{
          streakcount = 1;
        }
        count++;
      }
      results.push_back(count);
      count = 0;
      heads_streak = 0;
      tails_streak = 0;
      streakcount = 1;
    }
    int sum = accumulate(results.begin(), results.end(), 0);
    int average = sum / results.size();
    cout << "The average number of coin flips to reach a streak of " << streak;
    cout << " for either heads or tails with " << trials << " trials is: " << average << endl;

}
void coin_flip_heads_percentage(int flips){
  srand(time(0));
  int num_heads = 0;
  for (int flip = 0; flip < flips; flip++){
    if(rand() % 2 == 0){
      num_heads++;
    }
  }
  float percentage = (float(num_heads) /float(flips)) * 100;
  cout << "With " << flips << " coin tosses, " << setprecision(4) << percentage;
  cout << " percent of the flips landed on heads." << endl;
}

void random_number_of_the_day(int min, int max){
  srand(time(0));
  cout << "Random number of the day from " << min <<" to "<< max << ": ";
  cout << (rand() % 100) + 1 << endl;
}

int main(){
  roll_dice(2);
  coin_flip(10, 100000);
  coin_flip_specific_average(10, 2020);
  dice_roll_side_streak(6, 6);
  dice_roll_streak_any(5, 100);
  coin_flip_streak_any(15, 100);
  coin_flip_heads_percentage(12345);
  random_number_of_the_day(1, 100);
  return 0;
}
